### ISSUE TITLE
[Proposal] Distributed connection state and guaranteed message delivery

### ISSUE THREAD

**Issue by mkosieradzki** [aspnet/SignalR#242]
This is related to: #223 

Currently ConnectionManager does not allow to be initiated after service restart from some different source and also does not allow backing connection state to a different store.

Let's consider following scenario:
1. Service Fabric-based Stateful Service which is hosting a SignalR Hub.
2. Clients connect to this Hub using AspNetCore Proxy hosted by a Front facing Stateless Service.
3. Logical state of all SignalR Connections is persisted to a ReliableDictionary and replicated to secondary replicas.
4. This Stateful Service does suddenly a failover to a different cluster node (Resource Load Balancer decision or just some hardware failure).
5. Underlying transport connections are broken and clients try to reconnect.
6. New Primary Replica is ready to accept connections.
7. Logical SignalR connections are reestablished.

Next step would be to add some delivery guarantees and message buffering in a replicated ReliableQueue.

Does it sound like something considered for implementing in SignalR or is it something that should be implemented in a layer above SignalR?

I am not sure if this is in or out of the scope of the project...

https://aspnet.uservoice.com/forums/228522-show-me-how-with-code/suggestions/6206997-signalr-guaranteed-message-delivery

This uservoice says it might be actually in-scope...

If so, I would be happy to help with development of such features. However message delivery guarantee has some serious impact on the protocol.

If not, connection state failover has not as big value as it could...

**Comment by davidfowl** [comment]
Something like this won't be built into SignalR but we are interested in having the right extensibility points so that this is possible. Things are still pretty early with new SignalR but it's much closer to a programming model of pure websockets than the prior version. That said, I would start with the redis provider to get an idea of the current extensibility:

https://github.com/aspnet/SignalR/blob/af5e7da4786a41d3c0179f1d0be137e60b349b15/src/Microsoft.AspNetCore.SignalR.Redis/RedisHubLifetimeManager.cs

Also, the way we've been tackling the extensibility points is by writing the manual wire up code end users would have to write (in various technologies) to better understand what support needs to be added. If we can't even write that code, then we're in a bad place (that was one of the big downsides of existing SignalR).

**Comment by mkosieradzki** [comment]
@davidfowl Thanks, I have already browsed through the extensibility points code and I know the Redis example very well. (And already have browsed through nearly entire current SignalR core code).

I think, thanks to your explanation, now I get it:
1. One layer of SignalR is that it brings up an abstraction over WebSockets and different transports with web sockets-like programming model (Microsoft.AspNetCore.Sockets).
2. Second layer allows you to implement hubs using HubLifetimeManager any way you want as long it compliant with Hub-protocol:
However if you want to implement proper deliver guarantees while continuing to use you need to add two things to the protocol:
a) InvocationDescriptor should contain some serial number - to allow detecting duplicate invocations (on the client side!)
b) And some ACK message type
I understand that might be not interesting - but if I wanted to build an alternative Hub implementation with guaranteed delivery I can build on top of the Microsoft.AspNetCore.Sockets. I also would need to reimplement the client to respond to acks and ignore duplicate messages  - fair enough.


From my analysis (as stated before) Microsoft.AspNetCore.Sockets layer is missing one important extensibility point. ConnectionManager should be somehow replaceable to allow storing connection state in distributed collections.

At first, sticking only to the Microsoft.AspNetCore.Sockets layer, I will try to create a pluggable implementation of ReliableConnectionManager (based on ReliableCollections) and try to identify the best extensibility points to the ConnectionManager, if you are OK with it.

Later I can try creating some proposals for Hub-protocol implementation to allow guaranteed delivery, but only if this is potentially in the project scope.

**Comment by davidfowl** [comment]
> From my analysis (as stated before) Microsoft.AspNetCore.Sockets layer is missing one important extensibility point. ConnectionManager should be somehow replaceable to allow storing connection state in distributed collections.

The connection manager's job is to track connections on a specific instance. You need something else that sits above sockets. That's why endpoints get access to the entire connection object. There's a strict layering difference between the live connection and the "ReliableConnection" in this world, and one doesn't replace the other. Again, think of the connection as a socket, they come and go, but the reliable connection can exist or many connections.

> Later I can try creating some proposals for Hub-protocol implementation to allow guaranteed delivery, but only if this is potentially in the project scope.

We can look at what it would take to support ACKs in the protocol so you wouldn't need to write a new client.

/cc @DamianEdwards @anurse

**Comment by mkosieradzki** [comment]
@davidfowl If I understand correctly ConnectionManager handles the "ReliableConnections" this is why it has the TryGetConnection method which is used for the reconnection purpose.

ConnectionState = ReliableConnection

or am I missing something?

**Comment by mkosieradzki** [comment]
OK I think I get it. Sorry it's all OK. It should work. If I understand correctly if someone reconnects to a different node, their connection has the same ID. So there is no need to replicate any state.

**Comment by davidfowl** [comment]
@mkosieradzki There's no reconnect support in this version of SignalR. Long polling is the only transport that supports it (but that's just to function). When the client disconnects, that connection id is lost. ConnectionID in this world is tied to a specific node. You should have a reliable connection id that would be used to track distributed connections across nodes.

**Comment by mkosieradzki** [comment]
@davidfowl OK. This explains why I got so confused. I have read tons of documentation from legacy versions about reconnects etc.

**Comment by davidfowl** [comment]
@mkosieradzki see https://vimeo.com/204078084

**Comment by mkosieradzki** [comment]
@davidfowl Thanks a lot, I will definitely watch it!

**Comment by mkosieradzki** [comment]
@davidfowl That's a really great presentation. I like the changes very much and I think that the new model is better fit to the Stateful Service scenarios than the old one.

Regarding ACK: I don't think it is really needed as all method invocations eventually returns. Also the serial number can be passed by the first parameter to avoid duplication and this can be tracked in the application layer. So there is no need for changes.

Regarding reliable connections: they also can and should be implemented in the HubLifetimeManager.

Being aware there is no more such thing as reconnection. I will attempt to implement #223, especially the aforementioned stateful service failover one.

Thanks again for all the explantions. I am closing this one.

**Comment by davidfowl** [comment]
@mkosieradzki did you ever do anything here?

**Comment by mkosieradzki** [comment]
@davidfowl Nope. I have agreed that simplistic approach you have implemented is actually much better (at least in my scenario). Again: your explanation were very helpful.

Regarding #223 I've decided to build custom protocol based on protobuf on top of ASP.NET Core Sockets (from my perspective the biggest part of SignalR Core value-proposition is inside ASP.NET Core Sockets) connecting to a single stateful service replica in Service Fabric.

However, I have postponed my actual implementation until ASP.NET Core Sockets are ready - due to extremely low quality of WebSockets implementation across the entire ASP.NET Core ecosystem. There were tons of bugs in most trivial things like aspnet/AspNetCoreModule#77 and no one bothered fixing them until SignalR Core started getting more attention.
